### PR TITLE
Guard offset container allocation overflow

### PR DESCRIPTION
**Pull request by jkwak-work**
Automated review PR created by /review-on-fork-repo skill.

## Summary
- Adds overflow guards in `OffsetContainer` allocation paths to prevent integer overflow when computing new buffer sizes.
- Extends the offset container unit test to cover the new overflow handling.

### CODE REVIEW

**Comment by jkwak-work** [comment]
@coderabbitai review

**Comment by jkwak-work** [comment]
@greptileai

File: source/core/slang-offset-container.cpp [hunk 1]
```diff
@@ -150,19 +165,28 @@ void* OffsetContainer::allocate(size_t size, size_t alignment)
         size_t newSize = (calcSize < minSize) ? minSize : calcSize;
 
         // Reallocate space
-        m_data = (uint8_t*)::realloc(m_data, newSize);
+        uint8_t* newData = (uint8_t*)::realloc(m_data, newSize);
+        if (!newData)
+        {
+            return nullptr;
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp:184]: Addressed in commit 507594ed. All callers (newObject, newArray, newString) now check the return value of allocate() and propagate the failure as a null Offset32Ptr / empty Offset32Array.

File: source/core/slang-offset-container.cpp [hunk 2]
```diff
@@ -129,12 +129,27 @@ void OffsetContainer::fixAlignment(size_t alignment)
...
 {
-    size_t offset = (m_dataSize + alignment - 1) & ~(alignment - 1);
+    if (alignment == 0)
+    {
+        return nullptr;
+    }
 
-    if (offset + size > m_capacity)
+    const size_t alignmentMask = alignment - 1;
+    if (m_dataSize > SIZE_MAX - alignmentMask)
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp]: Addressed in commit 507594ed. The bound is now 0xFFFFFFFFu (introduced as kMaxDataSize). This enforces the 32-bit offset range that Offset32Ptr can represent and prevents silent truncation in getOffset().

File: source/core/slang-offset-container.cpp [hunk 3]
```diff
@@ -129,12 +129,27 @@ void OffsetContainer::fixAlignment(size_t alignment)
...
     {
-        const size_t minSize = offset + size;
+        return nullptr;
+    }
+
+    size_t offset = (m_dataSize + alignmentMask) & ~alignmentMask;
+    if (size > SIZE_MAX - offset)
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp]: Addressed in commit 507594ed. The size check on line 144 now uses kMaxDataSize (0xFFFFFFFFu) — same 32-bit bound as the m_dataSize check above.

File: source/core/slang-offset-container.cpp [hunk 4]
```diff
@@ -129,12 +129,27 @@ void OffsetContainer::fixAlignment(size_t alignment)
 
 void* OffsetContainer::allocate(size_t size, size_t alignment)
 {
-    size_t offset = (m_dataSize + alignment - 1) & ~(alignment - 1);
+    if (alignment == 0)
+    {
+        return nullptr;
+    }
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp]: Addressed in commit 507594ed. The first check in allocate() now also rejects non-power-of-two alignments via (alignment & (alignment - 1)) != 0.

File: tools/slang-unit-test/unit-test-offset-container.cpp [hunk 5]
```diff
@@ -39,6 +50,9 @@ SLANG_UNIT_TEST(offsetContainer)
         _checkEncodeDecode(uint32_t(i));
     }
 
+    _checkAllocateOverflowDoesNotWrap(SIZE_MAX - 7, 16, 1);
+    _checkAllocateOverflowDoesNotWrap(SIZE_MAX - 3, 1, 8);
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-offset-container.cpp:54]: Addressed in commit 507594ed. Added 4GB-boundary cases (m_dataSize set just under 0xFFFFFFFFu, plus a 0x80000000 + 0x80000000 sum-overflow case), and a few non-power-of-two alignment cases for completeness.

File: source/core/slang-offset-container.cpp [hunk 6]
```diff
@@ -129,12 +129,27 @@ void OffsetContainer::fixAlignment(size_t alignment)
...
     {
-        const size_t minSize = offset + size;
+        return nullptr;
+    }
+
+    size_t offset = (m_dataSize + alignmentMask) & ~alignmentMask;
+    if (size > SIZE_MAX - offset)
+    {
+        return nullptr;
+    }
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp]: Addressed in commit 507594ed. newArray now guards against count*sizeof(T) overflow via size > SIZE_MAX / sizeof(T), and clamps count to the 32-bit range stored in Offset32Array. newString rejects stringSizes that do not fit in 32 bits before computing headSize + stringSize + 1.

File: source/core/slang-offset-container.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp]: Addressed in commit 507594ed (same fix as the earlier duplicate thread): allocate() now rejects non-power-of-two alignments via (alignment & (alignment - 1)) != 0 before the bitwise mask is computed.

File: source/core/slang-offset-container.cpp [hunk 7]
```diff
@@ -129,12 +129,33 @@ void OffsetContainer::fixAlignment(size_t alignment)
...
+    const size_t alignmentMask = alignment - 1;
+    if (m_dataSize > kMaxDataSize - alignmentMask)
     {
-        const size_t minSize = offset + size;
+        return nullptr;
+    }
 
+    size_t offset = (m_dataSize + alignmentMask) & ~alignmentMask;
+    if (size > kMaxDataSize - offset)
+    {
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp:158]: Addressed in commit f101e09c. Added a guard `if (alignmentMask > kMaxDataSize) return nullptr;` before the m_dataSize subtraction, so a power-of-two alignment > 2^32 on a 64-bit host is rejected before alignmentMask underflows the kMaxDataSize - alignmentMask expression. Also added a unit test (gated to 64-bit via `SIZE_MAX > 0xFFFFFFFFu`) that exercises alignments of 2^33 and 2^40.

File: source/core/slang-offset-container.cpp [hunk 8]
```diff
@@ -171,11 +201,23 @@ Offset32Ptr<OffsetString> OffsetContainer::newString(const UnownedStringSlice& s
...
+    if (stringSize > size_t(0xFFFFFFFFu))
+    {
+        return Offset32Ptr<OffsetString>();
+    }
+
     uint8_t head[OffsetString::kMaxSizeEncodeSize];
     size_t headSize = OffsetString::calcEncodedSize(stringSize, head);
 
     size_t allocSize = headSize + stringSize + 1;
     uint8_t* bytes = (uint8_t*)allocate(allocSize);
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/core/slang-offset-container.cpp:225]: Addressed in commit f101e09c. Tightened the stringSize bound to `size_t(0xFFFFFFFFu) - OffsetString::kMaxSizeEncodeSize - 1`, which both keeps the encoded string within the 32-bit offset domain and prevents `headSize + stringSize + 1` from wrapping on 32-bit hosts where size_t == uint32_t.
